### ISSUE TITLE
[BUG] Cron Expression causes workflow to fire twice in a row

### ISSUE THREAD

**Issue by MaxBrooks114** [elsa-workflows/elsa-core#5499]
## Description
I have a workflow that is triggered by a cron expression to run once a day. For some reason it causes the workflow to be triggered twice immediately. The workflow then never completes and stays in the suspended state even though all the activities have been completed successfully. 
## Steps to Reproduce

1. **Detailed Steps**: 
     var cronTrigger = new Cron
     {
         CronExpression = new("0 30 16 * * ?"),
         CanStartWorkflow = true
     };

4. **Reproduction Rate**: 
Happens every time. 

## Expected Behavior
Workflow triggers once and completes.

## Actual Behavior
workflow triggers twice and enters suspended state.

## Screenshots
![image](https://github.com/elsa-workflows/elsa-core/assets/45081361/2ef40f19-f781-4315-ba4e-6fd54d6f5330)

## Environment
Elsa 3.1.3, .net 8 on windows

## Additional Context
it is a flowchart workflow.

**Comment by MaxBrooks114** [comment]
The suspended state is unrelated. However the duplicate triggering still happens, unknown why.

**Comment by sfmskywalker** [comment]
@MaxBrooks114 @yutao815  Can you try with the latest 3.2.0-preview packages to see if the issue still exists?

**Comment by sfmskywalker** [comment]
I just tried to reproduce this issue, but it works without issue. I will close this as unreproducible, but if you can reproduce the issue with the latest preview packages in a sample project, then please feel free to attach a zip file and I'll check it out. Thank you.

**Comment by MaxBrooks114** [comment]
@sfmskywalker I can confirm this bug still happens for me on preview version 3.2.0-sonar-cloud.1502. I will try to reproduce with a sample project when i have time

**Comment by sfmskywalker** [comment]
3.2.0-sonar-cloud.1502 Is quite an old build. You might want to try 3.2.0-preview.xxx.

**Comment by MaxBrooks114** [comment]
@sfmskywalker same issue with 3.2.0-preview.1717 too.

**Comment by sfmskywalker** [comment]
Ok, thank you for checking.  I look forward to the repro.

**Comment by MaxBrooks114** [comment]
[ElsaCronBugReplication.zip](https://github.com/user-attachments/files/15778068/ElsaCronBugReplication.zip)


Here is the bug replicated - it is using the latest 3.3 preview as well. I am using the quartz integration heartbeat workflow sample as a base, and then added a custom activity. It should be noted that your sample as well as mine do not actually use the quartz scheduler or elsa.quartz paackage, only elsa.scheduling. 

I have replicated the bug using a specific type of cron expression similar to the ones where i encountered the bug originally,  namely a once a day cron expression. Oddly, not only can I reproduce the bug where the job is being fired twice simultaneously, but also repeats every second like the original incarnation of the workflow, like some sort of merging of the two corn expressions. 
![image](https://github.com/elsa-workflows/elsa-core/assets/45081361/3b5d4ea2-e812-4d42-b6da-3b8cc26c34eb)

as you can see in the screenshot of the code the cron expression should only be firing once, but in the console below we have heartbeat style logs, with two every second

**Comment by sfmskywalker** [comment]
Thank you for the sample project, I'll take a look soon.

**Comment by sfmskywalker** [comment]
I can confirm the issue; specifically, it happens the first time when triggers don't yet exist in the database.

**Comment by 4lexKislitsyn** [comment]
I believe `CreateSchedulesHostedService` cause this behaviour.  

`DefaultWorkflowDefinitionStorePopulator` call `GetWorkflowsAsync` of CLR Provider. Later `IndexTriggersAsync` method is used to **store triggers** and send notifications. `ScheduleWorkflows` class handles notification and calls `ScheduleAsync` method of `ITriggerScheduler`. After all this steps `CreateSchedulesHostedService` is trying to get all stored triggers and schedule it again.

Because `IScheduler` is scoped, it can't detect duplicates. If  `IScheduler`, `CronosCronParser` and `Func<IServiceProvider, ICronParser>` are registered as singleton scheduler will detect duplicates and replace triggers.

Changes to SchedullingFeature.cs:
```csharp
            .AddSingleton<IScheduler, LocalScheduler>()
            ...
            .AddSingleton<CronosCronParser>()
            .AddSingleton(CronParser)
```

**Comment by MaxBrooks114** [comment]
@4lexKislitsyn I can confirm this fixed the issue, but you need to make ISystemClock feature a singleton as well. Should I make a pr or will you?

@sfmskywalker please confirm this can be a possible solution.

**Comment by 4lexKislitsyn** [comment]
@MaxBrooks114 I can create a PR after @sfmskywalker confirm that this solution will not breake something else. Are there other features that should be revised beside ISystemClock?

**Comment by MaxBrooks114** [comment]
@4lexKislitsyn not as far as I'm aware, it worked fine with your suggested fixes and the systemclock.

**Comment by 4lexKislitsyn** [comment]
I looked at systemclock feature and it's already registered as singleton in `SystemClockFeature.cs`.

Also `IServiceProvider` in `LocalScheduler` is used for creating of `ScheduledCronTask`, `ScheduledRecurringTask`, `ScheduledSpecificInstantTask` and this classes have `IServiceScopeFactory` to initialize objects with scoped or transient lifetime. So I think there will not be any issue with it.

I will open PR in few minutes.

**Comment by dwoldo** [comment]
Is this a dupe? https://github.com/elsa-workflows/elsa-core/issues/5600

**Comment by sfmskywalker** [comment]
@dwoldo Yes, but it also mentions another issue where unpublishing a workflow does not seem to unschedule the timer teigger.

**Comment by sfmskywalker** [comment]
Fixed via e5ba9a8a9d31541ad24fa0ea0853cce5c4cc9601
